grants_tagger/tune_threshold.py

```diff
@@ -8,94 +8,199 @@
-            max_f1 = f1
-            optimal_threshold_i = candidate_threshold_i
-    return optimal_threshold_i, max_f1
+        new_tp = tp + candidate_tp - label_tp
+        new_fp = fp + candidate_fp - label_fp
+        new_fn = fn + candidate_fn - label_fn
+        new_f1 = new_tp / (new_tp + (new_fp + new_fn) / 2)
 
+        if new_f1 > max_f1:
+            max_f1 = new_f1
+            optimal_threshold = candidate_threshold
+            logger.debug(
+                f"Label: {label_i:4d} - f1: {max_f1:.6f} - Th: {optimal_threshold:.3f}"
+            )
+
+    return optimal_threshold
+
+
+def optimise_threshold(Y_test, Y_pred_proba, nb_thresholds=None):
+    # start with lowest posible threshold per label
+    optimal_thresholds = Y_pred_proba.min(axis=0)
+
+    if issparse(optimal_thresholds):
+        optimal_thresholds = np.array(optimal_thresholds.todense())
+
+    optimal_thresholds = optimal_thresholds.ravel()
 
-def optimise_threshold(Y_test, Y_pred_proba, nb_thresholds=None, init_threshold=None):
-    if init_threshold:
-        optimal_thresholds = [init_threshold] * Y_pred_proba.shape[1]
-    else:
-        # start with lowest posible threshold per label
-        optimal_thresholds = Y_pred_proba.min(axis=0).ravel()
     updated = True
 
+    # Convert to CSC for fast column retrieval when asking for labels
+    if issparse(Y_pred_proba):
+        Y_pred_proba = Y_pred_proba.tocsc()
     Y_pred = Y_pred_proba > optimal_thresholds
+    if issparse(Y_test):
+        Y_test = Y_test.tocsc()
+        Y_pred = csc_matrix(Y_pred)
+
     optimal_f1 = f1_score(Y_test, Y_pred, average="micro")
-    print("---Starting f1---")
+    cm = multilabel_confusion_matrix(Y_test, Y_pred)
+    cm = cm.sum(axis=0).ravel()
+    print("---Min f1---")
     print(f"{optimal_f1:.3f}\n")
 
     iterations = 0
     while updated:
-        print(f"---Iteration {iterations}---")
+        start = time.time()
+        logger.debug(f"---Iteration {iterations}---")
         updated = False
-        for label_i in range(Y_test.shape[1]):
-            # find threshold for label that maximises overall f1
-            optimal_threshold_i, max_f1 = argmaxf1(
-                optimal_thresholds,
-                Y_test,
-                Y_pred_proba,
-                label_i,
-                nb_thresholds,
-                iterations,
-            )
-            if max_f1 > optimal_f1:
-                print(
-                    f"Label: {label_i:4d} - f1: {max_f1:.6f} - Th: {optimal_threshold_i:.3f}"
-                )
-                optimal_f1 = max_f1
-                optimal_thresholds[label_i] = optimal_threshold_i
+
+        nb_labels = Y_test.shape[1]
+        argmaxf1_partial = partial(
+            argmaxf1,
+            Y_test,
+            Y_pred_proba,
+            optimal_thresholds,
+            cm=cm,
+            nb_thresholds=nb_thresholds,
+            iterations=iterations,
+        )
+        with multiprocessing.Pool() as pool:
```

Comment:
`argmaxf1` seems like a great function - looking forward to the blog post with more details on the maths :)

Reply:
to be honest I think I will do that after Jan along with creating an open source package.